Operator in CrashLoop with panic

We are trying to switch to operator and after image update got to point of operator crashlooping:

```
2022-09-30T11:08:17.926Z        INFO    controller-runtime.manager.controller.solrcloud Starting EventSource    {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud", "source": "kind source: /, Ki
nd="}
2022-09-30T11:08:17.927Z        INFO    controller-runtime.manager.controller.solrcloud Starting Controller     {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud"}
2022-09-30T11:08:17.927Z        INFO    controller-runtime.manager.controller.solrcloud Starting workers        {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud", "worker count": 1}
2022-09-30T11:08:18.335Z        INFO    controller-runtime.manager.controller.solrprometheusexporter    Starting workers        {"reconciler group": "solr.apache.org", "reconciler kind": "SolrPrometheusExporter",
"worker count": 1}
2022-09-30T11:08:18.430Z        INFO    controller-runtime.manager.controller.solrcloud Update required because field changed   {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud", "name": "sear
ch-solr-test", "namespace": "search", "statefulSet": "search-solr-test-solrcloud", "kind": "statefulSet", "field": "Spec.Template.Spec.Volumes[1].VolumeSource", "from": {"secret":{"secretName":"gcp-search-solr-cre
dentials-secret","defaultMode":420}}, "to": {"secret":{"secretName":"gcp-search-solr-credentials-secret"}}}
2022-09-30T11:08:18.432Z        INFO    controller-runtime.manager.controller.solrcloud Updating StatefulSet    {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud", "name": "search-solr-test", "
namespace": "search", "statefulSet": "search-solr-test-solrcloud"}
2022-09-30T11:08:18.549Z        INFO    controller-runtime.manager.controller.solrcloud.ManagedUpdateSelector   Pod update selection started.   {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud
", "name": "search-solr-test", "namespace": "search", "outOfDatePods": 2, "maxPodsUnavailable": 1, "unavailableUpdatedPods": 0, "outOfDatePodsNotStarted": 0, "maxPodsToUpdate": 1}
2022-09-30T11:08:18.549Z        INFO    controller-runtime.manager.controller.solrcloud.ManagedUpdateSelector   Pod killed for update.  {"reconciler group": "solr.apache.org", "reconciler kind": "SolrCloud", "name
": "search-solr-test", "namespace": "search", "pod": "search-solr-test-solrcloud-0", "reason": "Pod's replicas are safe to take down, adhering to the minimum active replicas per shard."}
2022-09-30T11:08:18.549Z        INFO    controller-runtime.manager.controller.solrcloud.ManagedUpdateSelector   Pod update selection complete. Maximum number of pods able to be updated reached.       {"reconciler
group": "solr.apache.org", "reconciler kind": "SolrCloud", "name": "search-solr-test", "namespace": "search", "maxPodsToUpdate": 1}
E0930 11:08:18.550190       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 561 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x13dd140, 0x2249580})
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000760b00})
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:48 +0x75
panic({0x13dd140, 0x2249580})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/apache/solr-operator/controllers/util.EvictReplicasForPodIfNecessary({0x176d458, 0xc0058552c0}, 0xc005393b60, 0x1c, {0x1787d28, 0xc00547e6e0})
        /workspace/controllers/util/solr_update_util.go:493 +0x67
github.com/apache/solr-operator/controllers.(*SolrCloudReconciler).Reconcile(0xc0002b3e60, {0x176d458, 0xc0058552c0}, {{{0xc000885770, 0x144a920}, {0xc000885760, 0xc0008e4380}}})
        /workspace/controllers/solrcloud_controller.go:428 +0x3167
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000121180, {0x176d3b0, 0xc00029c000}, {0x1427120, 0xc000760b00})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298 +0x303
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000121180, {0x176d3b0, 0xc00029c000})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2({0x176d3b0, 0xc00029c000})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216 +0x46
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:213 +0x356
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x12b3627]

goroutine 561 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000760b00})
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x13dd140, 0x2249580})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/apache/solr-operator/controllers/util.EvictReplicasForPodIfNecessary({0x176d458, 0xc0058552c0}, 0xc005393b60, 0x1c, {0x1787d28, 0xc00547e6e0})
        /workspace/controllers/util/solr_update_util.go:493 +0x67
github.com/apache/solr-operator/controllers.(*SolrCloudReconciler).Reconcile(0xc0002b3e60, {0x176d458, 0xc0058552c0}, {{{0xc000885770, 0x144a920}, {0xc000885760, 0xc0008e4380}}})
        /workspace/controllers/solrcloud_controller.go:428 +0x3167
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000121180, {0x176d3b0, 0xc00029c000}, {0x1427120, 0xc000760b00})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298 +0x303
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000121180, {0x176d3b0, 0xc00029c000})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2({0x176d3b0, 0xc00029c000})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216 +0x46
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7f9eb86e9250)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0010c22a0, {0x1746200, 0xc000bfa4b0}, 0x1, 0xc0003103c0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0010ba6d0, 0x3b9aca00, 0x0, 0x0, 0xc0010c22d0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x176d3b0, 0xc00029c000}, 0xc0006677e0, 0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x176d3b0, 0xc00029c000}, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99 +0x2b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:213 +0x356

```

SolrCloud CRD definition:
```yaml
apiVersion: solr.apache.org/v1beta1
kind: SolrCloud
metadata:
  name: search-solr-test
  namespace: search
spec:
  busyBoxImage:
    repository: library/busybox
    tag: 1.28.0-glibc
  customSolrKubeOptions:
    podOptions:
      initContainers:
        - name: upload-zk-config
          image: xxx-docker.jfrog.io/xxx/search-solr:test-operator-3
          command: ["/var/solr-ricardo/scripts/load_configs_to_zookeeper.sh"]
          env:
            - name: ZK_HOST
              value: search-solr-test-solrcloud-zookeeper-0.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181,search-solr-test-solrcloud-zookeeper-1.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181,search-solr-test-solrcloud-zookeeper-2.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181/
      imagePullSecrets:
        - name: xxx-docker-jfrog
      envVars:
        - name: ARTICLES_EXPIRATION_FIELD
          value: expiration_date
        - name: AUTO_DELETE_PERIOD_SECONDS
          value: "3600"
        - name: GCS_PROJECT_ID
          value: xxxxx
        - name: GOOGLE_APPLICATION_CREDENTIALS
          value: "/var/gcp-credentials/credentials.json"
      volumes:
        - name: gcp-credentials
          defaultContainerMount:
            name: gcp-credentials
            mountPath: /var/gcp-credentials
            readOnly: true
          source:
            secret:
              secretName: gcp-search-solr-credentials-secret
  dataStorage:
    persistent:
      pvcTemplate:
        metadata:
          annotations:
            volume.beta.kubernetes.io/storage-class: regional-ssd
          name: search-solr-test
        spec:
          resources:
            requests:
              storage: 20Gi
  replicas: 3
  solrAddressability:
    commonServicePort: 80
    podPort: 8983
  solrImage:
    repository: xxx-docker.jfrog.io/xxx/search-solr
    tag: test-operator-3
  solrJavaMem: -Xms2048m -Xmx4096m
  solrLogLevel: INFO
  updateStrategy:
    managed: { }
    method: Managed
  zookeeperRef:
    provided:
      chroot: /
      config: { }
      image:
        pullPolicy: IfNotPresent
        repository: pravega/zookeeper
      replicas: 3
      zookeeperPodPolicy:
        resources: { }

```


Status on CRD:
```yaml
status:
  backupRestoreReady: false
  internalCommonAddress: http://search-solr-test-solrcloud-common.search
  podSelector: solr-cloud=search-solr-test,technology=solr-cloud
  readyReplicas: 2
  replicas: 3
  solrNodes:
  - internalAddress: http://search-solr-test-solrcloud-0.search-solr-test-solrcloud-headless.search:8983
    name: search-solr-test-solrcloud-0
    nodeName: gke-dev-cookie-e2-spoon-np-1fd433b7-l0z6
    ready: true
    specUpToDate: false
    version: test-operator
  - internalAddress: http://search-solr-test-solrcloud-1.search-solr-test-solrcloud-headless.search:8983
    name: search-solr-test-solrcloud-1
    nodeName: gke-dev-cookie-e2-fork-np-f4c2fe51-jkvi
    ready: true
    specUpToDate: false
    version: test-operator
  - internalAddress: http://search-solr-test-solrcloud-2.search-solr-test-solrcloud-headless.search:8983
    name: search-solr-test-solrcloud-2
    nodeName: gke-dev-cookie-e2-fork-np-3162ff26-q7vc
    ready: false
    specUpToDate: true
    version: test-operator-3
  targetVersion: test-operator-3
  upToDateNodes: 1
  version: test-operator
  zookeeperConnectionInfo:
    chroot: /
    externalConnectionString: N/A
    internalConnectionString: search-solr-test-solrcloud-zookeeper-0.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181,search-solr-test-solrcloud-zookeeper-1.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181,search-solr-test-solrcloud-zookeeper-2.search-solr-test-solrcloud-zookeeper-headless.search.svc.cluster.local:2181

```

Let me know if I can provide more info which can help you to investigate an issue.

--- a/api/v1beta1/solrbackup_types.go
+++ b/api/v1beta1/solrbackup_types.go
@@ -121,7 +121,7 @@ type IndividualSolrBackupStatus struct {
 	// +optional
 	CollectionBackupStatuses []CollectionBackupStatus `json:"collectionBackupStatuses,omitempty"`
 
-	// Version of the Solr being backed up
+	// The time that this backup was finished
 	// +optional
 	FinishTime *metav1.Time `json:"finishTimestamp,omitempty"`
 

--- a/api/v1beta1/solrcloud_types.go
+++ b/api/v1beta1/solrcloud_types.go
@@ -844,8 +844,7 @@ type ZookeeperSpec struct {
 	// +optional
 	ReadOnlyACL *ZookeeperACL `json:"readOnlyAcl,omitempty"`
 
-	// ZooKeeper ACL to use when connecting with ZK for reading operations.
-	// This ACL should have READ permission in the given chRoot.
+	// Additional Zookeeper Configuration settings
 	// +optional
 	Config ZookeeperConfig `json:"config,omitempty"`
 }
@@ -1416,6 +1415,16 @@ func (sc *SolrCloud) AdvertisedNodeHost(nodeName string) string {
 
 func (sc *SolrCloud) UsesPersistentStorage() bool {
 	return sc.Spec.StorageOptions.PersistentStorage != nil
+}
+
+func (sc *SolrCloud) DataVolumeName() string {
+	// Set the default name of the pvc
+	if sc.Spec.StorageOptions.PersistentStorage != nil &&
+		sc.Spec.StorageOptions.PersistentStorage.PersistentVolumeClaimTemplate.ObjectMeta.Name != "" {
+		return sc.Spec.StorageOptions.PersistentStorage.PersistentVolumeClaimTemplate.ObjectMeta.Name
+	} else {
+		return "data"
+	}
 }
 
 func (sc *SolrCloud) SharedLabels() map[string]string {

--- a/config/crd/bases/solr.apache.org_solrbackups.yaml
+++ b/config/crd/bases/solr.apache.org_solrbackups.yaml
@@ -153,7 +153,7 @@ spec:
                   type: object
                 type: array
               finishTimestamp:
-                description: Version of the Solr being backed up
+                description: The time that this backup was finished
                 format: date-time
                 type: string
               finished:
@@ -200,7 +200,7 @@ spec:
                         type: object
                       type: array
                     finishTimestamp:
-                      description: Version of the Solr being backed up
+                      description: The time that this backup was finished
                       format: date-time
                       type: string
                     finished:

--- a/config/crd/bases/solr.apache.org_solrclouds.yaml
+++ b/config/crd/bases/solr.apache.org_solrclouds.yaml
@@ -5267,7 +5267,7 @@ spec:
                         description: The ChRoot to connect solr at
                         type: string
                       config:
-                        description: ZooKeeper ACL to use when connecting with ZK for reading operations. This ACL should have READ permission in the given chRoot.
+                        description: Additional Zookeeper Configuration settings
                         properties:
                           additionalConfig:
                             additionalProperties:

--- a/controllers/util/solr_update_util.go
+++ b/controllers/util/solr_update_util.go
@@ -482,15 +482,16 @@ func GetAllManagedSolrNodeNames(solrCloud *solr.SolrCloud) map[string]bool {
 // This function MUST be idempotent and return the same list of pods given the same kubernetes/solr state.
 func EvictReplicasForPodIfNecessary(ctx context.Context, solrCloud *solr.SolrCloud, pod *corev1.Pod, logger logr.Logger) (err error, canDeletePod bool) {
 	var solrDataVolume *corev1.Volume
+	dataVolumeName := solrCloud.DataVolumeName()
 	for _, volume := range pod.Spec.Volumes {
-		if volume.Name == SolrDataVolumeName {
+		if volume.Name == dataVolumeName {
 			solrDataVolume = &volume
 			break
 		}
 	}
 
 	// Only evict if the Data volume is not persistent
-	if solrDataVolume.VolumeSource.PersistentVolumeClaim == nil {
+	if solrDataVolume != nil && solrDataVolume.VolumeSource.PersistentVolumeClaim == nil {
 		// If the Cloud has 1 or zero pods, and this is the "-0" pod, then delete the data since we can't move it anywhere else
 		// Otherwise, move the replicas to other pods
 		if (solrCloud.Spec.Replicas == nil || *solrCloud.Spec.Replicas < 2) && strings.HasSuffix(pod.Name, "-0") {

--- a/controllers/util/solr_util.go
+++ b/controllers/util/solr_util.go
@@ -45,7 +45,6 @@ const (
 	SolrCloudPVCTechnology           = "solr-cloud"
 	SolrPVCStorageLabel              = "solr.apache.org/storage"
 	SolrCloudPVCDataStorage          = "data"
-	SolrDataVolumeName               = "data"
 	SolrPVCInstanceLabel             = "solr.apache.org/instance"
 	SolrXmlMd5Annotation             = "solr.apache.org/solrXmlMd5"
 	SolrXmlFile                      = "solr.xml"
@@ -143,18 +142,14 @@ func GenerateStatefulSet(solrCloud *solr.SolrCloud, solrCloudStatus *solr.SolrCl
 		},
 	}
 
-	solrDataVolumeName := SolrDataVolumeName
+	solrDataVolumeName := solrCloud.DataVolumeName()
 
 	var pvcs []corev1.PersistentVolumeClaim
 	if solrCloud.UsesPersistentStorage() {
 		pvc := solrCloud.Spec.StorageOptions.PersistentStorage.PersistentVolumeClaimTemplate.DeepCopy()
 
 		// Set the default name of the pvc
-		if pvc.ObjectMeta.Name == "" {
-			pvc.ObjectMeta.Name = solrDataVolumeName
-		} else {
-			solrDataVolumeName = pvc.ObjectMeta.Name
-		}
+		pvc.ObjectMeta.Name = solrDataVolumeName
 
 		// Set some defaults in the PVC Spec
 		if len(pvc.Spec.AccessModes) == 0 {

--- a/helm/solr-operator/Chart.yaml
+++ b/helm/solr-operator/Chart.yaml
@@ -53,16 +53,13 @@ annotations:
   # Add change log for a single release here.
   # Allowed syntax is described at: https://artifacthub.io/docs/topics/annotations/helm/#example
   artifacthub.io/changes: |
-    - kind: added
-      description: Addition 1
+    - kind: fixed
+      description: Fix bug with named PVCs
       links:
         - name: Github Issue
-          url: https://github.com/issue-url
-    - kind: changed
-      description: Change 2
-      links:
+          url: https://github.com/apache/solr-operator/issues/479
         - name: Github PR
-          url: https://github.com/pr-url
+          url: https://github.com/apache/solr-operator/pull/481
   artifacthub.io/images: |
     - name: solr-operator
       image: apache/solr-operator:v0.7.0-prerelease

--- a/helm/solr-operator/crds/crds.yaml
+++ b/helm/solr-operator/crds/crds.yaml
@@ -153,7 +153,7 @@ spec:
                   type: object
                 type: array
               finishTimestamp:
-                description: Version of the Solr being backed up
+                description: The time that this backup was finished
                 format: date-time
                 type: string
               finished:
@@ -200,7 +200,7 @@ spec:
                         type: object
                       type: array
                     finishTimestamp:
-                      description: Version of the Solr being backed up
+                      description: The time that this backup was finished
                       format: date-time
                       type: string
                     finished:
@@ -5499,7 +5499,7 @@ spec:
                         description: The ChRoot to connect solr at
                         type: string
                       config:
-                        description: ZooKeeper ACL to use when connecting with ZK for reading operations. This ACL should have READ permission in the given chRoot.
+                        description: Additional Zookeeper Configuration settings
                         properties:
                           additionalConfig:
                             additionalProperties:

